In [1]:
#Importing aal the necessary libraries
import pandas as pd
import numpy as np

In [2]:
#Displaying all the columns
pd.set_option('display.max_columns',None)

In [3]:
#Loading the dataset
df = pd.read_csv("data.csv")
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St    2.0    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St    2.0    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St    2.0    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St    3.0    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St    3.0    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
0  03-09-2016       2.5    3067.0       2.0       1.0  1.0     126.0   
1  03-12-2016       2.5    3067.0       2.0       1.0  1.0     202.0   
2  04-02-2016       2.5    3067.0       2.0       1.0  0.0     156.0   
3  04-02-2016       2.5    3067.0       3.0       2.0  1.0       0.0   
4  04-03-2017       2.5    3067.0       3.0       2.0  0.0     134.0   

   BuildingArea  YearBuilt CouncilArea  Lattitude  Longtitude  \
0           NaN        NaN       Yarra   -37.8014    144.9958   
1           NaN        NaN       Yarra   -37.7996    144.9984   
2          79.0     1900.0       Yarra   -37.8079    144.9934   
3           NaN        NaN       Yarra   -37.8114    145.0116   
4         150.0     1900.0       Yarra   -37.8093    144.9944   

              Regionname  Propertycount  
0  Northern Metropolitan         4019.0  
1  Northern Metropolitan         4019.0  
2  Northern Metropolitan         4019.0  
3  Northern Metropolitan         4019.0  
4  Northern Metropolitan         4019.0

### 1. Determine the no. of rows and no. of columns in the dataset.

In [4]:
#Checking the shape
df.shape

(23678, 21)

### 2. Display the datatypes of each column.

In [5]:
#Checking the data type
df.dtypes

Suburb            object
Address           object
Rooms            float64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

### 3. Determine the row indexes of rows having all null values (i.e. empty rows).

In [6]:
null_index = df.loc[df.isnull().sum(axis=1)>=21].index

In [7]:
null_index

Int64Index([   31,  1145,  1348,  2119,  3277,  3288,  3745,  3752,  3798,
             3855,
            ...
            21220, 21343, 21602, 22003, 22109, 22288, 22704, 22918, 23153,
            23341],
           dtype='int64', length=131)

In [8]:
null_index.tolist()

[31,
 1145,
 1348,
 2119,
 3277,
 3288,
 3745,
 3752,
 3798,
 3855,
 4089,
 4096,
 4164,
 4218,
 4222,
 4326,
 4649,
 4897,
 4942,
 5052,
 5223,
 5265,
 5328,
 5331,
 5383,
 5405,
 5625,
 5741,
 5952,
 6125,
 6133,
 6140,
 6143,
 6251,
 6326,
 6406,
 6446,
 6703,
 6717,
 6853,
 6921,
 7086,
 7261,
 7348,
 7596,
 7743,
 7759,
 7762,
 7775,
 8054,
 8095,
 8118,
 8190,
 8409,
 8501,
 8652,
 8816,
 8974,
 9098,
 9210,
 9232,
 9450,
 9455,
 9586,
 9761,
 9888,
 9894,
 10009,
 10125,
 10390,
 10475,
 10738,
 10952,
 11334,
 11348,
 11552,
 11755,
 12087,
 12441,
 12559,
 12997,
 13066,
 13381,
 13647,
 13725,
 14009,
 14471,
 14644,
 14963,
 15094,
 15214,
 15771,
 16044,
 16252,
 16563,
 16655,
 16848,
 17111,
 17246,
 17611,
 17741,
 18077,
 18148,
 18251,
 18374,
 18476,
 18547,
 18671,
 18742,
 18884,
 19001,
 19287,
 19641,
 19949,
 20087,
 20422,
 20496,
 20695,
 20838,
 20880,
 21055,
 21220,
 21343,
 21602,
 22003,
 22109,
 22288,
 22704,
 22918,
 23153,
 23341]

In [9]:
#Checking the length 
len(null_index)

131

Observation:
    
    131 rows having all values null.

### 4. Determine the total number of empty rows and their percentage in whole dataset and remove them.

In [11]:
#Creating the copy of our dataset
df1 = df.copy()

In [12]:
df.shape

(23678, 21)

In [14]:
df.dropna(how='all').shape

(23547, 21)

In [61]:
#Checking the total number of empty rows
23678-23547

131

In [17]:
#Empty rows percentage
((23678-23547)/23678)*100

0.5532561871779711

Finding:
    0.55% empty rows in the datasets.

In [18]:
#Dropping the rows
df.dropna(how='all',inplace=True)

In [19]:
df.shape

(23547, 21)

### 5. Determine the column names having missing values.

In [20]:
#Columns having missing values
df.columns[df.isnull().any()]

Index(['Price', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

### 6. Determine the missing values and their percentage in each column.

In [21]:
df.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

In [22]:
(df.isnull().sum()/ len(df)) * 100

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.875398
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.004247
Postcode          0.004247
Bedroom2         19.030025
Bathroom         19.042766
Car              19.645815
Landsize         26.062768
BuildingArea     57.455302
YearBuilt        50.991634
CouncilArea      33.511700
Lattitude        18.278337
Longtitude       18.278337
Regionname        0.004247
Propertycount     0.004247
dtype: float64

### 7. Determine the row indexes having missing values.

In [23]:
missing_rows= df.loc[df.isnull().sum(axis=1)>0].index

In [24]:
#missing_rows.tolist()
missing_rows

Int64Index([    0,     1,     3,     5,     7,     8,     9,    10,    12,
               13,
            ...
            23668, 23669, 23670, 23671, 23672, 23673, 23674, 23675, 23676,
            23677],
           dtype='int64', length=17351)

### 8. Determine the missing values and their percentage in each row.

In [25]:
missing_values = df.loc[df.isnull().sum(axis=1)>0].index

In [26]:
(missing_values / len(df)).tolist()

[0.0,
 4.2468254979402894e-05,
 0.00012740476493820868,
 0.00021234127489701447,
 0.0002972777848558203,
 0.00033974603983522316,
 0.0003822142948146261,
 0.00042468254979402894,
 0.0005096190597528347,
 0.0005520873147322376,
 0.0006370238246910434,
 0.0006794920796704463,
 0.0007219603346498493,
 0.000806896844608655,
 0.0008493650995880579,
 0.0008918333545674609,
 0.0009343016095468637,
 0.0009767698645262667,
 0.0011041746294644752,
 0.0011466428844438782,
 0.0011891111394232812,
 0.001231579394402684,
 0.0013589841593408926,
 0.0014439206692996986,
 0.0014863889242791013,
 0.001571325434237907,
 0.0016987301991761158,
 0.0017411984541555188,
 0.0017836667091349217,
 0.0019535397290525334,
 0.001996007984031936,
 0.002038476239011339,
 0.002080944493990742,
 0.002165881003949548,
 0.0022508175139083534,
 0.0022932857688877564,
 0.0023357540238671593,
 0.0023782222788465623,
 0.0025905635537435768,
 0.0026755000637023823,
 0.0027604365736611882,
 0.002802904828640591,
 0.0029727778

### 9. Determine the column names having more than 30% missing values and remove them from the dataset.

In [27]:
df.columns[df.isnull().sum()/len(df)*100 > 30]

Index(['BuildingArea', 'YearBuilt', 'CouncilArea'], dtype='object')

In [28]:
df.drop(columns=['BuildingArea','YearBuilt','CouncilArea'], axis=1,inplace = True)

In [29]:
df.shape

(23547, 18)

### 10. Determine the number and percentage of rows having more than 5 missing values. If the percentage of such rows is less than 20% then remove all these rows.

In [30]:
missing_row = df.loc[df.isnull().sum(axis=1)>5].index

In [31]:
missing_row

Int64Index([   15,    16,    19,    21,    22,    27,    34,    35,    46,
               47,
            ...
            23573, 23577, 23580, 23584, 23603, 23626, 23630, 23633, 23639,
            23662],
           dtype='int64', length=4278)

In [32]:
#Converting it into percentage
4278/len(df)

0.1816791948018856

Conclusion:

1. Row index having more than 5 null value is 4278.
2. Row having more than 5 null value persent is 18% and less than 20% so it drop.

In [36]:
#Dropping the rows
df.drop(df[df.isnull().sum(axis=1)>5].index,inplace=True)

In [37]:
df.shape

(19269, 18)

### 11. Display the description of data and percentage of missing values still left column wise.

In [38]:
#Description of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19269 entries, 0 to 23677
Data columns (total 18 columns):
Suburb           19269 non-null object
Address          19269 non-null object
Rooms            19269 non-null float64
Type             19269 non-null object
Price            15086 non-null float64
Method           19269 non-null object
SellerG          19269 non-null object
Date             19269 non-null object
Distance         19269 non-null float64
Postcode         19269 non-null float64
Bedroom2         19066 non-null float64
Bathroom         19063 non-null float64
Car              18921 non-null float64
Landsize         17410 non-null float64
Lattitude        19243 non-null float64
Longtitude       19243 non-null float64
Regionname       19269 non-null object
Propertycount    19269 non-null float64
dtypes: float64(11), object(7)
memory usage: 2.8+ MB


In [39]:
#Percentage of the missing values
(df.isnull().sum()/len(df))*100

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.708444
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.000000
Postcode          0.000000
Bedroom2          1.053506
Bathroom          1.069075
Car               1.806010
Landsize          9.647621
Lattitude         0.134932
Longtitude        0.134932
Regionname        0.000000
Propertycount     0.000000
dtype: float64

### 12. Considering the 'Price' column as target variable which is to be predicted, how will you treat the missing values? Give a brief explanation and treat the missing values in 'Price' column.

In [40]:
df['Price'].isnull().sum()

4183

In [41]:
(df['Price'].isnull().sum()/len(df)) * 100

21.70844361409518

Explanation:
    Price column is a target column and in this column more than 20% null value are present.So we are not dropping all the null
    value of Price columns because if we drop all the null value of the price columns more data is lost so drop more data 
    it's not good way.

In [42]:
#price column missing value fillup by mean 
df['Price'].fillna(df['Price'].mean(),inplace=True)

### 13. Now check the percentage of missing values in each column again, if missing values exist, treat it with the most suitable method.

In [43]:
#Percentage of the missing values
(df.isnull().sum()/len(df))*100

Suburb           0.000000
Address          0.000000
Rooms            0.000000
Type             0.000000
Price            0.000000
Method           0.000000
SellerG          0.000000
Date             0.000000
Distance         0.000000
Postcode         0.000000
Bedroom2         1.053506
Bathroom         1.069075
Car              1.806010
Landsize         9.647621
Lattitude        0.134932
Longtitude       0.134932
Regionname       0.000000
Propertycount    0.000000
dtype: float64

In [44]:
df.columns[df.isnull().any()]

Index(['Bedroom2', 'Bathroom', 'Car', 'Landsize', 'Lattitude', 'Longtitude'], dtype='object')

In [45]:
#Imputing the missing value with mode 

for col in ['Bedroom2', 'Bathroom', 'Car', 'Landsize', 'Lattitude', 'Longtitude']:
    df[col].fillna(df[col].mean(), inplace=True)

In [51]:
#Checking the missing values in data set
df.isnull().values.any()

False

Observation:
    There is no null values present in our dataset.

### 14. Now since you have completed your data cleaning, determine the number and percentage of columns removed.

In [52]:
df.shape

(19269, 18)

In [53]:
df1.shape

(23678, 21)

Conclusion:
    Three columns are removed.

In [54]:
#Percentage of columns
(3/21)*100

14.285714285714285

### 15. Determine the number and percentage of rows removed in whole process. Also comment briefly that according to you the amount of data lost is justified or not.

In [55]:
len(df)

19269

In [56]:
len(df1)

23678

In [57]:
#Number of rows removed
len(df1) - len(df)

4409

In [60]:
#Percentage of rows removed
((len(df1) - len(df)) / len(df1)) * 100

18.620660528760872

### 4409 rows and 18.62% of rows removed in whole data cleaning process. If data lost is more than 7 to 10% then it is not good. But in this dataset we remove 18% of our data which might causes some information loss.